src/actions/login.js

```diff
@@ -1,65 +1,129 @@
 import * as types from '../constants/actionTypes';
 import firebase from 'firebase';
+import base from '../constants/base';
+import oauth from 'panoptes-client/lib/oauth';
 
 // References for our Firebase listener
 let userListener;
 let userRef;
 
-export function checkLoginUser() {
-  return (dispatch) => {
-    firebase.auth().onAuthStateChanged(user => {
-      user
-        ? dispatch(setLoginUser(user))
-        : console.log('User not logged in')
-    })
-  }
-}
 
 function setLoginUser(user) {
   return dispatch => {
-    console.info('Logged in as', user.providerData[0].displayName)
     dispatch({
       type: types.USER_LOGIN,
-      payload: user
-    })
-    userRef = firebase.database().ref(`users/${user.uid}`);
-    userListener = userRef.on('value', dataSnapshot => {
-      console.info('Updating userVotes object...')
-      const voteData = dataSnapshot.child('votes').val();
-      dispatch({
-        type: types.USERVOTES_ADD,
-        payload: voteData
-      });
+      payload: user,
     });
-  }
+    if (user) {
+      userRef = firebase.database().ref(`users/${user.uid}`);
+      userListener = userRef.on('value', dataSnapshot => {
+        console.info('Updating userVotes object...');
+        const voteData = dataSnapshot.child('votes').val();
+        dispatch({
+          type: types.USERVOTES_ADD,
+          payload: voteData,
+        });
+      });
+    }
+  };
 }
 
-export function login() {
+
+function getFirebaseToken(token) {
+//  don't hard code the endpoint
+  return fetch('http://localhost:8080/validate?token=' + token, {
+    method: 'GET',
+    mode: 'cors',
+    headers: new Headers({
+      'Content-Type': 'application/json',
+    }),
+  })
+  .then(response => response.json())
+  .then(json => json.token)
+  .catch(error => console.error('ERROR: ', error));
+}
+
+function firebaseLogin(apiToken) {
   return (dispatch) => {
-    const provider = new firebase.auth.GithubAuthProvider();
-    return firebase.auth().signInWithPopup(provider)
-      .then(data => {
-        // Check if we have an error object back
-        if (data.code) {
-          Promise.reject(data);
+    getFirebaseToken(apiToken)
+      .then(firebaseToken => {
+        if (firebaseToken) {
+          firebase.auth().signInWithCustomToken(firebaseToken)
+            .then(userData => {
+              dispatch(setLoginUser(userData));
+              console.log('Firebase login successful.');
+            })
+            .catch(error => {
+              console.log('error: ', error);
+            });
         } else {
-          dispatch(setLoginUser(data.user))
+          console.error('Undefined Firebase token');
         }
-      })
-      .catch(error => {
-        console.error('Error logging in: ', error);
       });
-  }
+  };
+}
+
+function setFirebaseUserDisplayname(panoptesDisplayName) {
+  firebase.auth().onAuthStateChanged((user) => {
```

Comment:
I'm not sure this is right - we shouldn't create a listener for something that only happens once on login